src/Tests/library/test_rwlocks.pl

```diff
@@ -0,0 +1,112 @@
+/*  Part of SWI-Prolog
+
+    Author:        Younes A
+    E-mail:        dev@younes.io
+
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions
+    are met:
+
+    1. Redistributions of source code must retain the above copyright
+       notice, this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above copyright
+       notice, this list of conditions and the following disclaimer in
+       the documentation and/or other materials provided with the
+       distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+:- asserta(user:file_search_path(library, './library')).
+:- use_module(library(rwlocks)).
+
+:- begin_tests(rwlocks).                                                                                                                                          
+                                                                                                                                                                      
+    test(priority_handling) :-
+        retractall(started(_)),
+        with_rwlock(test_lock, 
+                    (assertz(started(write)), 
+                    sleep(1), 
+                    writeln('Write operation completed')), 
+                    write),
+        with_rwlock(test_lock, 
+                    (assertz(started(read)), 
+                    sleep(2), 
+                    writeln('Read operation completed')), 
+                    read),
+        findall(X, started(X), Started),
+        writeln(Started),
+        assertion(Started == [write, read]).                                                                                                                    
+                                                                                                                                                                      
+    test(timeout_handling, [throws(time_limit_exceeded(rwlock))]) :-
+        thread_create(with_rwlock(test_lock, sleep(2), write), WriteThread, []),
+        sleep(0.1),  % Give some time for the write lock to be acquired
```

Comment:
Sleeping to sync always always causes failing tests at some point.   Instead, use a thread message to sync.   It makes the test run faster (99% of the time) and reliable, even on heavily loaded or emulated machines.